text/stacks/0007-add-database-libs.md

```diff
@@ -0,0 +1,188 @@
+# Add database libraries to Full stack
+
+## Summary
+
+Many users of Paketo buildpacks build applications that connect to databases.
+In order to facilitate the onboarding experience of these users into the Paketo
+ecosystem, we should add packages to the Full stack that provide the necessary
+header files for applications to connect to databases.
+
+Specifically, we should add `libpq-dev` and `libmariadb-dev` to the Full stack.
+
+## Motivation
+
+We want to make it easy for application developers to try out Paketo buildpacks
+to build applications that connect to databases. This is a common pattern, and
+one that we should support out-of-the-box.
+
+Currently in order to build such an application, consumers must also first
+build their own stack. This is a significant increase in effort and may cause
+application developers to rule out Paketo Buildpacks as an option for building
+their applications.
+
+Additionally, for customers who are willing to use the Full stack as-is and
+who value timely security updates, building their own stack also incurs the
+ongoing maintenance burden of keeping this stack up to date.
+
+The Full stack is positioned to optimize onboarding experience for most typical
+applications, at the expense of file size and security surface area. It is
+reasonable that this definition of "typical applications" includes applications
+that connect to databases.
+
+These database libraries have many dependencies (see below for more details) so
+we must balance the improved user experience with the increased security risk
+from more packages. Adding the libraries to the Full stack offers an acceptable
+compromise.
+
+## Detailed Explanation
+
+For the Ubuntu 18.04 (Bionic) Full stack and Ubuntu 22.04 (Jammy) Full stacks,
+we should add the following packages to the `stack.toml` files:
+
+* libmariadb-dev
+* libpq-dev
```

Comment:
When you say  "in the stack", do you mean in the run image? Today, both of these packages are available in the full build image, but not the run image. Is the request to add these to the run image?

Reply:
They are in the build image? That's interesting. Rob created this RFC based on a question from myself. I'm packaging a RoR app wich has mysql2 gem:
```
gem 'mysql', '>= 0.3.18', '< 0.5'
```

The build fails when trying to install that gem with the following error, which I interpret as lack of mysql-dev libs:

```
[builder]       Installing mysql2 0.4.10 with native extensions
[builder]       Gem::Ext::BuildError: ERROR: Failed to build gem native extension.
[builder]       
[builder]       current directory:
[builder]       /layers/paketo-buildpacks_bundle-install/build-gems/gems/mysql2-0.4.10/ext/mysql2
[builder]       /layers/paketo-buildpacks_mri/mri/bin/ruby -I
[builder]       /layers/paketo-buildpacks_mri/mri/lib/ruby/2.7.0 -r
[builder]       ./siteconf20221102-70-45wmtq.rb extconf.rb
[builder]       checking for rb_absint_size()... yes
[builder]       checking for rb_absint_singlebit_p()... yes
[builder]       checking for ruby/thread.h... yes
[builder]       checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
[builder]       checking for rb_thread_blocking_region()... no
[builder]       checking for rb_wait_for_single_fd()... yes
[builder]       checking for rb_hash_dup()... yes
[builder]       checking for rb_intern3()... yes
[builder]       checking for rb_big_cmp()... yes
[builder]       *** extconf.rb failed ***
[builder]       Could not create Makefile due to some reason, probably lack of necessary
[builder]       libraries and/or headers.  Check the mkmf.log file for more details.  You may
[builder]       need configuration options.
[builder]       
[builder]       Provided configuration options:
[builder]       	--with-opt-dir
[builder]       	--without-opt-dir
[builder]       	--with-opt-include
[builder]       	--without-opt-include=${opt-dir}/include
[builder]       	--with-opt-lib
[builder]       	--without-opt-lib=${opt-dir}/lib
[builder]       	--with-make-prog
[builder]       	--without-make-prog
[builder]       	--srcdir=.
[builder]       	--curdir
[builder]       	--ruby=/layers/paketo-buildpacks_mri/mri/bin/$(RUBY_BASE_NAME)
[builder]       	--with-ruby-dir
[builder]       	--without-ruby-dir
[builder]       	--with-ruby-include
[builder]       	--without-ruby-include=${ruby-dir}/include
[builder]       	--with-ruby-lib
[builder]       	--without-ruby-lib=${ruby-dir}/lib
[builder]       	--with-mysql-dir
[builder]       	--without-mysql-dir
[builder]       	--with-mysql-include
[builder]       	--without-mysql-include=${mysql-dir}/include
[builder]       	--with-mysql-lib
[builder]       	--without-mysql-lib=${mysql-dir}/lib
[builder]       	--with-mysql-config
[builder]       	--without-mysql-config
[builder]       	--with-mysqlclient-dir
[builder]       	--without-mysqlclient-dir
[builder]       	--with-mysqlclient-include
[builder]       	--without-mysqlclient-include=${mysqlclient-dir}/include
[builder]       	--with-mysqlclient-lib
[builder]       	--without-mysqlclient-lib=${mysqlclient-dir}/lib
[builder]       	--with-mysqlclientlib
[builder]       	--without-mysqlclientlib
[builder]       /layers/paketo-buildpacks_mri/mri/lib/ruby/2.7.0/mkmf.rb:1050:in `block in
[builder]       find_library': undefined method `split' for nil:NilClass (NoMethodError)
[builder]       	from /layers/paketo-buildpacks_mri/mri/lib/ruby/2.7.0/mkmf.rb:1050:in `collect'
[builder]       from /layers/paketo-buildpacks_mri/mri/lib/ruby/2.7.0/mkmf.rb:1050:in
[builder]       `find_library'
[builder]       	from extconf.rb:95:in `<main>'
[builder]       
[builder]       To see why this extension failed to compile, please check the mkmf.log which can
[builder]       be found here:
[builder]       
[builder]       /layers/paketo-buildpacks_bundle-install/build-gems/extensions/x86_64-linux/2.7.0-static/mysql2-0.4.10/mkmf.log
[builder]       
[builder]       extconf failed, exit code 1
[builder]       
[builder]       Gem files will remain installed in
[builder]       /layers/paketo-buildpacks_bundle-install/build-gems/gems/mysql2-0.4.10 for
[builder]       inspection.
[builder]       Results logged to
[builder]       /layers/paketo-buildpacks_bundle-install/build-gems/extensions/x86_64-linux/2.7.0-static/mysql2-0.4.10/gem_make.out
[builder]       
[builder]       An error occurred while installing mysql2 (0.4.10), and Bundler cannot continue.
[builder]       Make sure that `gem install mysql2 -v '0.4.10' --source 'https://rubygems.org/'`
[builder]       succeeds before bundling.
[builder]       
[builder]       In Gemfile:
[builder]         mysql2
[builder] failed to execute bundle install output:
[builder] 
[builder] error: exit status 5
[builder] ERROR: failed to build: exit status 1
ERROR: failed to build: executing lifecycle. This may be the result of using an untrusted builder: failed with status code: 51
```

People are reporting that [installing of `default-libmysqlclient-dev` will help](https://github.com/brianmario/mysql2/issues/1160).

Reply:
> When you say "in the stack", do you mean in the run image? Today, both of these packages are available in the full build image, but not the run image. Is the request to add these to the run image?

I meant build image - as these libraries shouldn't be needed at runtime - but I was looking in the wrong section of the stack descriptor file and didn't see that they already exist in the build image 😳 

However, I think we need to investigate the bug mentioned by @cz4rny above.

I'll move this RFC to draft while we investigate which libraries are needed, and I'll re-open it if we determine that the full stack should have any additionally libraries on the build image.

Reply:
@cz4rny I opened this [issue](https://github.com/paketo-buildpacks/ruby/issues/862) to track the investigation.

Reply:
Great, thanks @robdimsdale !